code/iaas/api-logic/src/main/java/io/cattle/platform/iaas/api/auth/impl/GithubOAuthImpl.java

```diff
@@ -36,31 +47,58 @@ public Account getAccount(ApiRequest request) {
         }
-        String accountId = validateAndFetchAccountId(token);
-        return authDao.getAccountByExternalId(accountId, GITHUB_ACCOUNT_TYPE);
-
-    }
+        String accountId = githubUtils.validateAndFetchAccountIdFromToken(token);
+        if (null == accountId) {
+            return null;
+        }
+        Account account = authDao.getAccountByExternalId(accountId, GITHUB_ACCOUNT_TYPE);
 
-    private String validateAndFetchAccountId(String token) {
-        String toParse = null;
-        String[] tokenArr = token.split("\\s+");
-        if (tokenArr.length == 2) {
-            if(!StringUtils.equals("bearer", StringUtils.lowerCase(StringUtils.trim(tokenArr[0])))) {
-                return null;
+        String projectId = request.getServletContext().getRequest().getHeader(PROJECT_HEADER);
+        if (StringUtils.isEmpty(projectId)) {
+            projectId = request.getServletContext().getRequest().getParameter("projectId");
+            if (StringUtils.isEmpty(projectId)) {
+                return account;
             }
-            toParse = tokenArr[1];
-        } else if (tokenArr.length == 1) {
-            toParse = tokenArr[0];
-        } else {
-            return null;
         }
-        Map<String, Object> jsonData;
+
+        String unobfuscatedId = null;
+
         try {
-            jsonData = tokenService.getJsonPayload(toParse, true);
-        } catch (TokenException e) { //in case of invalid token
+            unobfuscatedId = ApiContext.getContext().getIdFormatter().parseId(projectId);
+        } catch (NumberFormatException e) {
+            throw new ClientVisibleException(ResponseCodes.BAD_REQUEST, "InvalidFormat", "projectId header format is incorrect " + projectId, null);
+        }
+
+        if (StringUtils.isEmpty(unobfuscatedId)) {
             return null;
         }
-        return (String) jsonData.get("account_id");
+
+        Account projectAccount = authDao.getAccountById(new Long(unobfuscatedId));
+
+        if (validateProjectAccount(projectAccount, token)) {
+            return projectAccount;
+        }
+        return null;
+    }
+
+    protected boolean validateProjectAccount(Account projectAccount, String token) {
+        if (null == projectAccount) {
+            return false;
+        }
+        Account account = projectAccount;
+        List<String> accesibleIds = null;
+        if (StringUtils.equals(TEAM_SCOPE, account.getExternalIdType())) {
+            accesibleIds = githubUtils.validateAndFetchTeamIdsFromToken(token);
+        } else if (StringUtils.equals(ORG_SCOPE, account.getExternalIdType())) {
+            accesibleIds = githubUtils.validateAndFetchOrgIdsFromToken(token);
+        }
+        if (StringUtils.equals(USER_SCOPE, projectAccount.getExternalIdType())) {
+            String unobfuscatedUserId = ApiContext.getContext().getIdFormatter().parseId(projectAccount.getExternalId());
+            account = authDao.getAccountById(new Long(unobfuscatedUserId));
```

Comment:
In case the scope of the project is "user", then the external Id of the project is set to that user's accountId.